man/ghi.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "GHI" "1" "June 2016" "Stephen Celis" "GHI Manual"
.
.SH "NAME"
\fBghi\fR \- the stupid issue tracker
.
.SH "SYNOPSIS"
\fBghi\fR [\fB\-\-version\fR] [\fB\-p\fR|\fB\-\-paginate\fR|\fB\-\-no\-pager\fR] [\fB\-\-help\fR] \fIcommand\fR [\fIargs\fR]
.
.br
\fB\fR \fB\fR \fB\fR \fB\fR [ \fB\-\-\fR [\fIuser\fR/]\fIrepo\fR]]
.
.SH "DESCRIPTION"
GHI is the alphabetical acronym for GitHub Issues\. GHI provides an easy way to manage your issues from the comfort of the command line and your editor of choice\.
.
.P
GHI should look familiar if you\'re familiar with Git\. When invoked from a repository directory that has a GitHub remote, GHI will work with that repository\'s issues by default\. Otherwise, GHI will work with your global issues or with a repository specified in its terminating arguments\.
.
.P
The \fIcommand\fR is either the name of a GHI command (see below) or an alias\.
.
.SH "OPTIONS"
.
.TP
\fB\-\-version\fR
Prints the GHI version number\.
.
.TP
\fB\-\-help\fR
Prints the synopsis and a list of commonly\-used commands\. If a GHI command is named, this option will bring up the manual page for that command\.
.
.IP
See \fBghi\-help\fR(1) for more information\.
.
.TP
\fB\-p\fR, \fB\-\-paginate\fR
Pipe all output into less if standard output is a terminal\.
.
.TP
\fB\-\-no\-pager\fR
Do not pipe output into a pager\.
.
.TP
\fB\-\-[no\-]color\fR
Toggle colorized output\.
.
.SH "GHI COMMANDS"
.
.TP
\fBghi\-assign\fR(1)
Assign an issue to yourself (or someone else)\.
.
.TP
\fBghi\-close\fR(1)
Close an issue\.
.
.TP
\fBghi\-lock\fR(1)
Lock an issue\'s conversation, limiting it to collaborators\.
.
.TP
\fBghi\-unlock\fR(1)
Unlock an issue\'s conversation, opening it to all viewers\.
.
.TP
\fBghi\-comment\fR(1)
Leave a comment on an issue\.
.
.TP
\fBghi\-config\fR(1)
Configure GHI\.
.
.TP
\fBghi\-edit\fR(1)
Modify an existing issue\.
.
.TP
\fBghi\-label\fR(1)
Create, list, modify, or delete labels\.
.
.TP
\fBghi\-list\fR(1)
List your issues (or a repository\'s)\.
.
.TP
\fBghi\-milestone\fR(1)
Manage project milestones\.
.
.TP
\fBghi\-open\fR(1)
Open (or reopen) an issue\.
.
.TP
\fBghi\-show\fR(1)
Show an issue\'s details\.
.
.SH "ENVIRONMENT VARIABLES"
Various GHI commands use the following environment variables:
.
.TP
\fIGHI_REPO\fR
The GitHub repository to be used by the issued GHI command\.
.
.TP
\fIGHI_TOKEN\fR
The OAuth access token used to authenticate you\.
.
.TP
\fIGHI_PAGER\fR
This environment variable overrides \fI$GIT_PAGER\fR and \fI$PAGER\fR\. If it is set to an empty string or to the value "cat", git will not launch a pager\.
.
.TP
\fIGHI_EDITOR\fR
This environment variable overrides \fI$GIT_EDITOR\fR, \fI$VISUAL\fR, and \fI$EDITOR\fR\.
.
.TP
\fIGITHUB_USER\fR
Your GitHub username: used to authenticate and identify you\.
.
.TP
\fIGITHUB_PASSWORD\fR
If you store your GitHub password in an environment variable, GHI will automatically attempt to fetch an access token using it\.
.
.P
All environment variables may also be stored in Git configuration files where the variable is in lowercase form and the underscore is a period (e\.g\., \fI$GHI_REPO\fR would become the ghi\.repo option)\.
.
.SH "AUTHORS"
GHI is maintained by Stephen Celis\. A full list of contributors can be found at \fIhttps://github\.com/stephencelis/ghi/contributors\fR\.
.
.SH "REPORTING BUGS"
.
.nf

ghi open \-\- stephencelis/ghi
.
.fi
.
.P
Or visit \fIhttps://github\.com/stephencelis/ghi/issues\fR\.


lib/ghi.rb
require 'optparse'

module GHI
  autoload :Authorization, 'ghi/authorization'
  autoload :Client,        'ghi/client'
  autoload :Commands,      'ghi/commands'
  autoload :Editor,        'ghi/editor'
  autoload :Formatting,    'ghi/formatting'
  autoload :Web,           'ghi/web'

  IGNORE_MARKER = '#|#'

  class << self
    attr_reader :current_command

    def execute args
      STDOUT.sync = true
      @current_command = "#{$0} #{args.join(' ')}".freeze

      double_dash = args.index { |arg| arg == '--' }
      if index = args.index { |arg| arg !~ /^-/ }
        if double_dash.nil? || index < double_dash
          command_name = args.delete_at index
          command_args = args.slice! index, args.length
        end
      end
      command_args ||= []

      option_parser = OptionParser.new do |opts|
        opts.banner = <<EOF
usage: ghi [--version] [-p|--paginate|--no-pager] [--help] <command> [<args>]
           [ -- [<user>/]<repo>]
EOF
        opts.on('--version') { command_name = 'version' }
        opts.on '-p', '--paginate', '--[no-]pager' do |paginate|
          GHI::Formatting.paginate = paginate
        end
        opts.on '--help' do
          command_args.unshift(*args)
          command_args.unshift command_name if command_name
          args.clear
          command_name = 'help'
        end
        opts.on '--[no-]color' do |colorize|
          Formatting::Colors.colorize = colorize
        end
        opts.on '-l' do
          if command_name
            raise OptionParser::InvalidOption
          else
            command_name = 'list'
          end
        end
        opts.on '-v' do
          command_name ? self.v = true : command_name = 'version'
        end
        opts.on('-V') { command_name = 'version' }
      end

      begin
        option_parser.parse! args
      rescue OptionParser::InvalidOption => e
        warn e.message.capitalize
        abort option_parser.banner
      end

      if command_name.nil?
        command_name = 'list'
      end

      if command_name == 'help'
        Commands::Help.execute command_args, option_parser.banner
      else
        command_name = fetch_alias command_name, command_args
        begin
          command = Commands.const_get command_name.capitalize
        rescue NameError
          abort "ghi: '#{command_name}' is not a ghi command. See 'ghi --help'."
        end

        # Post-command help option parsing.
        Commands::Help.execute [command_name] if command_args.first == '--help'

        begin
          command.execute command_args
        rescue OptionParser::ParseError, Commands::MissingArgument => e
          warn "#{e.message.capitalize}\n"
          abort command.new([]).options.to_s
        rescue Client::Error => e
          if e.response.is_a?(Net::HTTPNotFound) && Authorization.token.nil?
            raise Authorization::Required
          else
            abort e.message
          end
        rescue SocketError => e
          abort "Couldn't find internet."
        rescue Errno::ECONNREFUSED, Errno::ETIMEDOUT => e
          abort "Couldn't find GitHub."
        end
      end
    rescue Authorization::Required => e
      retry if Authorization.authorize!
      warn e.message
      if Authorization.token
        warn <<EOF.chomp

Not authorized for this action with your token. To regenerate a new token:
EOF
      end
      warn <<EOF

Please run 'ghi config --auth <username>'
EOF
      exit 1
    end

    def config key, options = {}
      upcase = options.fetch :upcase, true
      flags = options[:flags]
      var = key.gsub('core', 'git').gsub '.', '_'
      var.upcase! if upcase
      value = ENV[var] || `git config #{flags} #{key}`
      value = `#{value[1..-1]}` if value.start_with? '!'
      value = value.chomp
      value unless value.empty?
    end

    attr_accessor :v
    alias v? v

    private

    def fetch_alias(command, args)
      Commands::Aliases.fetch(command, args)
    end

  end
end
